File: pdlearning/src/microservices/course-microservice/Api/Application/Models/LearningPathModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microservice.Course.Domain.Entities;
using Microservice.Course.Domain.Enums;

namespace Microservice.Course.Application.Models
{
    public class LearningPathModel
    {
        public Guid Id { get; set; }

        public string Title { get; set; }

        public string ThumbnailUrl { get; set; }

        public DateTime? CreatedDate { get; set; }

        public Guid CreatedBy { get; set; }

        public LearningPathStatus Status { get; set; }

        public List<LearningPathCourseModel> ListCourses { get; set; }

        // Metadata
        public IEnumerable<string> CourseLevelIds { get; set; }

        public IEnumerable<string> PDAreaThemeIds { get; set; }

        public IEnumerable<string> ServiceSchemeIds { get; set; }

        public IEnumerable<string> SubjectAreaIds { get; set; }

        public IEnumerable<string> LearningFrameworkIds { get; set; }

        public IEnumerable<string> LearningDimensionIds { get; set; }

        public IEnumerable<string> LearningAreaIds { get; set; }

        public IEnumerable<string> LearningSubAreaIds { get; set; }

        public IEnumerable<string> MetadataKeys { get; set; }

        public static LearningPathModel Create(LearningPath learningPath, IEnumerable<LearningPathCourse> learningPathCourses)
        {
            return new LearningPathModel
            {
                Id = learningPath.Id,
                Title = learningPath.Title,
                ThumbnailUrl = learningPath.ThumbnailUrl,
                CreatedDate = learningPath.CreatedDate,
                CreatedBy = learningPath.CreatedBy,
                Status = learningPath.Status,
                ListCourses = learningPathCourses != null ? learningPathCourses.Select(p => LearningPathCourseModel.Create(p)).OrderBy(o => o.Order).ToList() : new List<LearningPathCourseModel>(),
                CourseLevelIds = learningPath.CourseLevelIds,
                PDAreaThemeIds = learningPath.PDAreaThemeIds,
                ServiceSchemeIds = learningPath.ServiceSchemeIds,
                SubjectAreaIds = learningPath.SubjectAreaIds,
                LearningFrameworkIds = learningPath.LearningFrameworkIds,
                LearningDimensionIds = learningPath.LearningDimensionIds,
                LearningAreaIds = learningPath.LearningAreaIds,
                LearningSubAreaIds = learningPath.LearningSubAreaIds,
                MetadataKeys = learningPath.MetadataKeys
            };
        }
    }
}

File: pdlearning/src/web-apps/src/libraries/domain-api/src/course/models/learning-path.model.ts
import { ILearningPathCourseModel, LearningPathCourseModel } from './learning-path-course-model';
import { SystemRoleEnum, UserInfoModel } from './../../share/models/user-info.model';

import { LMM_PERMISSIONS } from '@opal20/domain-api/share/permission-keys/lmm-permission-key';
export enum LearningPathStatus {
  Published = 'Published',
  Unpublished = 'Unpublished'
}

export interface ILearningPathModel {
  id?: string;
  title: string;
  createdDate: Date;
  createdBy?: string;
  status?: LearningPathStatus;
  thumbnailUrl?: string;
  listCourses: ILearningPathCourseModel[];
  description?: string;

  // Metadata
  courseLevelIds?: string[];
  pdAreaThemeIds?: string[];
  serviceSchemeIds?: string[];
  subjectAreaIds?: string[];
  learningFrameworkIds?: string[];
  learningDimensionIds?: string[];
  learningAreaIds?: string[];
  learningSubAreaIds?: string[];
  teacherOutcomeIds?: string[];
  categoryIds?: string[];
  teachingCourseStudyIds?: string[];
  teachingLevels?: string[];
  metadataKeys?: string[];
}

export class LearningPathModel implements ILearningPathModel {
  public id?: string;
  public title: string = '';
  public createdDate: Date = new Date();
  public createdBy?: string;
  public status?: LearningPathStatus = LearningPathStatus.Unpublished;
  public thumbnailUrl?: string;
  public listCourses: LearningPathCourseModel[] = [];

  // Metadata
  public courseLevelIds?: string[] = [];
  public pdAreaThemeIds?: string[] = [];
  public serviceSchemeIds: string[] = [];
  public subjectAreaIds: string[] = [];
  public learningFrameworkIds: string[] = [];
  public learningDimensionIds: string[] = [];
  public learningAreaIds: string[] = [];
  public learningSubAreaIds: string[] = [];
  public metadataKeys: string[] = [];

  public static hasCreateLearningPathPermission(user: UserInfoModel): boolean {
    return (
      user.hasPermissionPrefix(LMM_PERMISSIONS.CreateEditPublishUnpublishLP) ||
      user.hasAdministratorRoles() ||
      user.hasRole(SystemRoleEnum.CourseContentCreator)
    );
  }

  public static hasViewLearningPathDetailPermission(user: UserInfoModel): boolean {
    return user.hasPermissionPrefix(LMM_PERMISSIONS.ViewLearningPathDetail);
  }

  constructor(data?: ILearningPathModel) {
    if (data == null) {
      return;
    }
    this.id = data.id;
    this.title = data.title;
    this.createdDate = new Date(data.createdDate);
    this.createdBy = data.createdBy;
    this.thumbnailUrl = data.thumbnailUrl;
    this.listCourses = data.listCourses ? data.listCourses.map(p => new LearningPathCourseModel(p)) : [];
    this.status = data.status;

    // Metadata
    this.courseLevelIds = data.courseLevelIds;
    this.pdAreaThemeIds = data.pdAreaThemeIds;
    this.serviceSchemeIds = data.serviceSchemeIds;
    this.subjectAreaIds = data.subjectAreaIds;
    this.learningFrameworkIds = data.learningFrameworkIds;
    this.learningDimensionIds = data.learningDimensionIds;
    this.learningAreaIds = data.learningAreaIds;
    this.learningSubAreaIds = data.learningSubAreaIds;
    this.metadataKeys = data.metadataKeys;
  }

  public getAllMetadataTagIds(): string[] {
    return []
      .concat(this.courseLevelIds)
      .concat(this.pdAreaThemeIds)
      .concat(this.serviceSchemeIds)
      .concat(this.subjectAreaIds)
      .concat(this.learningFrameworkIds)
      .concat(this.learningDimensionIds)
      .concat(this.learningAreaIds)
      .concat(this.learningSubAreaIds);
  }

  public hasEditPublishUnpublishPermission(user: UserInfoModel): boolean {
    return user.hasPermissionPrefix(LMM_PERMISSIONS.CreateEditPublishUnpublishLP);
  }

  public hasViewCopyHyperLinkLearningPathButtonPermission(user: UserInfoModel): boolean {
    return user.hasPermissionPrefix(LMM_PERMISSIONS.CopyHyperLinkLearningPathButton);
  }

  public hasOwnerPermission(user: UserInfoModel): boolean {
    return this.createdBy === user.id || user.hasAdministratorRoles();
  }

  public hasPopulateMetadataLearningPathPermission(user: UserInfoModel): boolean {
    return (
      user.hasRole(SystemRoleEnum.CourseContentCreator) ||
      this.hasOwnerPermission(user) ||
      user.hasPermissionPrefix(LMM_PERMISSIONS.PopulateMetadataLearningPath)
    );
  }

  public isPublished(): boolean {
    return this.status === LearningPathStatus.Published;
  }
}
